src/system_database.ts

```diff
@@ -594,4 +605,56 @@ export class PostgresSystemDatabase implements SystemDatabase {
     };
     this.notificationsClient.on("notification", handler);
   }
+
+  /* SCHEDULER */
+  async getLastScheduledTime(wfn: string): Promise<number | null> {
+    const res = await this.pool.query<scheduler_state>(`
+      SELECT last_wf_sched_time
+      FROM ${DBOSExecutor.systemDBSchemaName}.scheduler_state
+      WHERE wf_function = $1;
+    `, [wfn]);
+
+    let v = res.rows[0]?.last_wf_sched_time ?? null;
+    if (v!== null) v = parseInt(`${v}`);
+    return v;
+  }
+
+  async setLastScheduledTime(wfn: string, invtime: number): Promise<number | null> {
+    const res = await this.pool.query<scheduler_state>(`
+      INSERT INTO ${DBOSExecutor.systemDBSchemaName}.scheduler_state (wf_function, last_wf_sched_time)
+      VALUES ($1, $2)
+      ON CONFLICT (wf_function)
+      DO UPDATE SET last_wf_sched_time = GREATEST(EXCLUDED.last_wf_sched_time, scheduler_state.last_wf_sched_time)
+      RETURNING last_wf_sched_time;
+    `, [wfn, invtime]);
+    
+    return parseInt(`${res.rows[0].last_wf_sched_time}`);
+  }
+
+  async getOutstandingScheduledWorkflows(wfn: string): Promise<number> {
+    const res = await this.pool.query<CountResult>(`
+      SELECT COUNT(scheduled_time) AS count
+      FROM ${DBOSExecutor.systemDBSchemaName}.scheduled_wf_running
+      WHERE wf_function = $1 AND scheduled_time IS NOT NULL;
+    `, [wfn]);
+    return parseInt(`${res.rows[0].count}`); // Returns the count of scheduled times
+  }
+  async scheduledWorkflowStarted(wfn: string, invtime: number): Promise<void> {
+    // if another node scheduled execution for this time slot, or it is a re-execute from a crashed run, this will do nothing
+    await this.pool.query(`
+      INSERT INTO ${DBOSExecutor.systemDBSchemaName}.scheduled_wf_running (wf_function, scheduled_time, actual_time)
+      VALUES ($1, $2, $3)
+      ON CONFLICT (wf_function, scheduled_time) DO NOTHING;
+    `, [wfn, `${invtime}`, `${new Date().getTime()}`]);
+  }
+  async scheduledWorkflowComplete(wfn: string, invtime: number): Promise<void> {
+    await this.pool.query(`
+      DELETE FROM ${DBOSExecutor.systemDBSchemaName}.scheduled_wf_running
+      WHERE wf_function = $1 AND scheduled_time = $2;
+    `, [wfn, `${invtime}`]);
+  }
```

Comment:
Looks like `scheduled_wf_running` is only used to track how many scheduled workflows are running. Do we need a separate table for this? Can we simply search `dbos.workflow_status` table:
```
select count(*) from dbos.workflow_status where name LIKE 'sched-<name>%' AND status = 'PENDING';
```

Reply:
I put a lot of thought into whether or not existing tables would be sufficient to establish this, without any timing holes, and thought it couldn't.

Imagine that the scenario is processing a batch of all recent records, and the scheduler is basically a polling loop.  Correctness is easily achieved if only one process is going in the system, more difficult otherwise.  Thus the interest in placing it as one would a semaphore - right before and after the code of interest.  Now, because of the OAOO key it is OK to take the semaphore if it is already held by a workflow with that key, because this must be a restart.  And then the restart may release it.

Reply:
I think the `workflow_status` table can establish this. When a workflow (non-temp) starts, we synchronously record the workflow into that table and set status to `PENDING` (this process also handles onConflict in a similar way); this is semaphore acquire. After a workflow finishes, we batch export the (successful) output, or we record error synchronously; this is semaphore release. If we crash before exporting output, then the workflow remains pending and will restart next time.

Another issue with the current implementation is that `getOutstandingScheduledWorkflows` doesn't give you consistent result if you restart a workflow, as it changes the input args after we return from recorded inputs. It would be an issue if a workflow with the same UUID gets different inputs across runs.